### PR TITLE
CLOUDSTACK-10007 Isolation methods

### PR DESCRIPTION
**Pull request by DaanHoogland**
changing isolation methods from an enum to a registry based construct to enhance plugability

### CODE REVIEW

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 996
 Hypervisor xenserver
 NetworkType Advanced
 Passed=105
 Failed=8
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 54 runs

* test_volumes.py

 * test_06_download_detached_volume Failed

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 54 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 87 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 87 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 85 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 85 runs

 * test_03_RVR_Network_check_router_state Failing since 2 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by DaanHoogland** [comment]
@fmaximus don't forget to rigorously look at this in terms of nuage's plugin ;)

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: api/src/com/cloud/network/PhysicalNetwork.java [hunk 1]
```diff
@@ -32,8 +35,82 @@
...
+            return Objects.hash(methodPrefix, provider);
+        }
+
+        @Override
+        public String toString() {
+            return methodPrefix.toString();
+        }
+
+        public static boolean remove(String prfx, String prvdr) {
+            if(prvdr == null || prvdr.isEmpty()) {
```
Comment (fmaximus) [inline api/src/com/cloud/network/PhysicalNetwork.java]: Better use the same logic as on line 81: `StringUtils.isNotBlank(prvdr)? prvdr : UNKNOWN_PROVIDER;`

File: api/src/com/cloud/network/PhysicalNetwork.java [hunk 2]
```diff
@@ -32,8 +35,82 @@
...
+                prvdr = UNKNOWN_PROVIDER;
+            }
+            return remove(new IsolationMethod(prfx, prvdr));
```
Comment (fmaximus) [inline api/src/com/cloud/network/PhysicalNetwork.java]: The constructor would first add it to the set, before removing it.
Would it be possible to use getIsolationMethod()?

File: api/src/com/cloud/network/PhysicalNetwork.java [hunk 3]
```diff
@@ -32,8 +35,82 @@
...
+         * gets a IsolationMethod object that defines this prefix and if any it returns the first one found that has a known provider. If none has a known provider
+         * it will return the one with the unknown provider. if none is found it return null.
+         *
+         * @param prfx
+         * @return
+         */
+        public static IsolationMethod getIsolationMethod(String prfx) {
+            IsolationMethod rc = null;
+            for (IsolationMethod method: registeredIsolationMethods) {
+                if (method.provider.equals(prfx)) {
```
Comment (DaanHoogland) [inline api/src/com/cloud/network/PhysicalNetwork.java]: good catch, should have passed any unit test :(

File: api/src/com/cloud/network/PhysicalNetwork.java [hunk 4]
```diff
@@ -32,8 +35,82 @@
...
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(methodPrefix, provider);
+        }
+
+        @Override
+        public String toString() {
+            return methodPrefix.toString();
```
Comment (DaanHoogland) [inline api/src/com/cloud/network/PhysicalNetwork.java]: sure

File: api/src/com/cloud/network/PhysicalNetwork.java (hunk 1 above)
Comment (DaanHoogland) [inline api/src/com/cloud/network/PhysicalNetwork.java]: don't care and will change

File: api/src/com/cloud/network/PhysicalNetwork.java (hunk 2 above)
Comment (DaanHoogland) [inline api/src/com/cloud/network/PhysicalNetwork.java]: writing an extra getIso...(prfx, String provider);

File: plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java [hunk 5]
```diff
@@ -133,7 +133,7 @@
 
     public NuageVspGuestNetworkGuru() {
         super();
-        _isolationMethods = new IsolationMethod[] {IsolationMethod.VSP};
+        _isolationMethods = new IsolationMethod[] {new IsolationMethod("VSP")};
```
Comment (DaanHoogland) [inline plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: I am not sure, but I think you are free to do so in your plugin, aren't you?

**Comment by DaanHoogland** [comment]
@fmaximus thanks for the review. good quality improvements arose ;)

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@DaanHoogland please put a CLOUDSTACK jira ID on it, squash your changes.

File: api/test/com/cloud/network/IsolationMethodTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+package com.cloud.network;
+
+import org.junit.After;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Created by dahn on 18/07/17.
```
Comment (rhtyd) [inline api/test/com/cloud/network/IsolationMethodTest.java]: Remove this comment.

File: api/src/com/cloud/network/PhysicalNetwork.java [hunk 7]
```diff
@@ -32,8 +36,101 @@
...
+        }
+
+        public IsolationMethod(String prfx, String prvdr) {
+            methodPrefix = prfx;
+            provider = StringUtils.isNotBlank(prvdr)? prvdr : UNKNOWN_PROVIDER;
+            registeredIsolationMethods.add(this);
+        }
+
+        /**
+         * gets a IsolationMethod object that defines this prefix and if any it returns the first one found that has a known provider. If none has a known provider
```
Comment (rhtyd) [inline api/src/com/cloud/network/PhysicalNetwork.java]: Remove comment, comment is explaining the code that a developer can easily read and follow.

File: plugins/network-elements/bigswitch-vns/src/com/cloud/network/guru/BigSwitchVnsGuestNetworkGuru.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+import org.apache.log4j.Logger;
+
+import javax.ejb.Local;
+import javax.inject.Inject;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+
+@Local(value = NetworkGuru.class)
+public class BigSwitchVnsGuestNetworkGuru extends GuestNetworkGuru {
```
Comment (rhtyd) [inline plugins/network-elements/bigswitch-vns/src/com/cloud/network/guru/BigSwitchVnsGuestNetworkGuru.java]: The PR says adding isolation type, why has this added a new network guru? Do we have h/w to test this network guru?

File: plugins/network-elements/nicira-nvp/src/com/cloud/network/guru/NiciraNvpGuestNetworkGuru.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+import org.apache.log4j.Logger;
+
+import javax.ejb.Local;
+import javax.inject.Inject;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+
+@Local(value = NetworkGuru.class)
+public class NiciraNvpGuestNetworkGuru extends GuestNetworkGuru {
```
Comment (rhtyd) [inline plugins/network-elements/nicira-nvp/src/com/cloud/network/guru/NiciraNvpGuestNetworkGuru.java]: Again, why this new class/network-guru implementation?

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 1013
 Hypervisor xenserver
 NetworkType Advanced
 Passed=107
 Failed=6
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 68 runs

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 68 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 101 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 101 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 99 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 99 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by DaanHoogland** [comment]
@rhtyd there is already a jira id on the title of the PR

File: plugins/network-elements/bigswitch-vns/src/com/cloud/network/guru/BigSwitchVnsGuestNetworkGuru.java (hunk 8 above)
Comment (DaanHoogland) [inline plugins/network-elements/bigswitch-vns/src/com/cloud/network/guru/BigSwitchVnsGuestNetworkGuru.java]: another rebase error, will remove it

**Comment by rhtyd** [comment]
@DaanHoogland the JIRA id in the PR is not enough and useful, if the JIRA id is part of the git commit (summary) it is easier to generate changelog (https://github.com/apache/cloudstack-docs-rn/tree/master/utils) and check/refer to PRs/issues later from git history. This has been an issue for me, and I found a workaround that is to use Github UI to squash/merge the PR and added the cloudstack ID (if specified in the PR summary).
Please fix conflicts and rebase against latest master.

File: plugins/network-elements/nicira-nvp/src/com/cloud/network/guru/NiciraNvpGuestNetworkGuru.java (hunk 9 above)
Comment (DaanHoogland) [inline plugins/network-elements/nicira-nvp/src/com/cloud/network/guru/NiciraNvpGuestNetworkGuru.java]: and another one

File: api/src/com/cloud/network/PhysicalNetwork.java (hunk 7 above)
Comment (DaanHoogland) [inline api/src/com/cloud/network/PhysicalNetwork.java]: no, not just developers will try and read the code. create a PR to remove it if it is in the way

**Comment by DaanHoogland** [comment]
@fmaximus can you give an ok for this?

**Comment by rhtyd** [comment]
Thanks @DaanHoogland LGTM.

**Pull request merged**